diff --git a/src/WebResponses.cpp b/src/WebResponses.cpp
index 22a549f..e0b36b3 100644
--- a/src/WebResponses.cpp
+++ b/src/WebResponses.cpp
@@ -318,7 +318,7 @@ size_t AsyncAbstractResponse::_ack(AsyncWebServerRequest *request, size_t len, u
           free(buf);
           return 0;
       }
-      outLen = sprintf_P((char*)buf+headLen, PSTR("%x"), readLen) + headLen;
+      outLen = sprintf_P((char*)buf+headLen, PSTR("%04x"), readLen) + headLen;
       while(outLen < headLen + 4) buf[outLen++] = ' ';
       buf[outLen++] = '\r';
       buf[outLen++] = '\n';
