diff --git a/RF24.cpp b/RF24.cpp
index 9e5b4a8..a4de63c 100644
--- a/RF24.cpp
+++ b/RF24.cpp
@@ -1871,6 +1871,11 @@ uint8_t RF24::getARC(void)
     return read_register(OBSERVE_TX) & 0x0F;
 }
 
+uint8_t RF24::getPLOS(void)
+{
+    return read_register(OBSERVE_TX) & 0x0F;
+}
+
 /****************************************************************************/
 
 bool RF24::setDataRate(rf24_datarate_e speed)
diff --git a/RF24.h b/RF24.h
index dbd32ae..a3d6b52 100644
--- a/RF24.h
+++ b/RF24.h
@@ -1644,6 +1644,7 @@ public:
      * @return Returns values from 0 to 15.
      */
     uint8_t getARC(void);
+    uint8_t getPLOS(void);
 
     /**
      * Set the transmission @ref Datarate
@@ -2415,4 +2416,4 @@ private:
  * Use `ctrl+c` to quit at any time.
  */
 
-#endif // __RF24_H__
\ No newline at end of file
+#endif // __RF24_H__
